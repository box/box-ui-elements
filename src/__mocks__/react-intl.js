/* eslint-disable import/no-extraneous-dependencies */
import * as React from 'react';
import PropTypes from 'prop-types';
import IntlMessageFormat from 'intl-messageformat';
import IntlRelativeTimeFormat from '@formatjs/intl-relativetimeformat';

export const FormattedDate = () => <div />;
FormattedDate.displayName = 'FormattedDate';

export const FormattedTime = () => <div />;
FormattedTime.displayName = 'FormattedTime';

export const FormattedRelativeTime = () => <div />;
FormattedRelativeTime.displayName = 'FormattedRelativeTime';

export const FormattedMessage = () => <div />;
FormattedMessage.displayName = 'FormattedMessage';

export const FormattedNumber = () => <div />;
FormattedNumber.displayName = 'FormattedNumber';

export const IntlProvider = () => <div />;
IntlProvider.displayName = 'IntlProvider';

export const defineMessages = messages => messages;
export const intlShape = PropTypes.any;

export const createIntl = ({ locale = 'en' }) => ({
    formatMessage: (message, values) => {
        const imf = new IntlMessageFormat(message.defaultMessage || message.message, locale);
        return imf.format(values);
    },
    formatDate: date => date,
    formatRelativeTime: (value, unit = 'second', options = {}) => {
        // eslint-disable-next-line
        IntlRelativeTimeFormat.__addLocaleData(
            // eslint-disable-next-line
            require(`@formatjs/intl-relativetimeformat/dist/locale-data/${locale}.json`),
        );
        const rtf = new IntlRelativeTimeFormat(locale, options);
        return rtf.format(value, unit);
    },
});

export const injectIntl = Component => {
    const WrapperComponent = props => {
        const injectedProps = {
            ...props,
            intl: createIntl({ locale: props.locale }), // eslint-disable-line react/prop-types
        };
        return <Component {...{ ...injectedProps }} />;
    };
    WrapperComponent.displayName = Component.displayName || Component.name || 'Component';
    return WrapperComponent;
};
