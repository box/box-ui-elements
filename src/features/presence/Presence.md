### Examples

```
const collaborators = [
    {
        avatarUrl: '',
        id: '1',
        name: 'A User',
        interactedAt: Date.now(),
        isActive: true
    },
    {
        avatarUrl: '',
        id: '2',
        name: 'B User',
        interactedAt: 1501300384000,
        interactionType: 'user.comment_create'
    },
    {
        avatarUrl: '',
        id: '3',
        name: 'C User',
        interactedAt: 1502216436000,
        interactionType: 'user.item_upload'
    },
    {
        avatarUrl: '',
        id: '4',
        name: 'D User',
        interactedAt: 1501527327000,
        interactionType: 'user.item_preview'
    },
    {
        avatarUrl: '',
        id: '5',
        name: 'E User',
        interactedAt: 1502217560000,
        interactionType: 'user.item_preview'
    }
];

const getLinkCallback = () => {};
const inviteCallback = () => {};
const avatarAttr = { 'data-resin-target': 'avatar' };
const containerAttr = { 'data-resin-feature': 'presence' };

<div style={ {"height": "300px"} }>
    <Presence
        collaborators={ collaborators }
        maxDisplayedAvatars={ 3 }
        getLinkCallback={ getLinkCallback }
        inviteCallback={ inviteCallback }
        avatarAttributes={ avatarAttr }
        containerAttributes={ containerAttr }
      />
</div>

```
