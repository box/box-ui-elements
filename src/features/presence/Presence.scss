@import './styles';

.presence {
    position: relative;
    box-shadow: 0 0 10px 0 rgba(0, 0, 0, .1);
}

.presence-avatar-container {
    position: relative;

    &.dropdown-active {
        .bdl-PresenceAvatarList-count {
            color: $bdl-gray-80;
            background-color: $bdl-gray-30;
        }
    }
}

.presence-link-container {
    position: relative;
    display: inline-block;
}

.presence-avatar {
    position: relative;
    display: inline-block;
}

.presence-avatar.is-active::after {
    position: absolute;
    right: 0;
    bottom: 2px;
    display: block;
    box-sizing: content-box;
    width: 6px;
    height: 6px;
    background-color: $bdl-green-light;
    border: 2px solid $white;
    border-radius: 50%;
    content: '';
}

.presence-avatar-notehead {
    margin-right: 2px;
}

.presence-avatar-dropdown {
    vertical-align: sub;
}

.presence-notehead {
    width: $bdl-Presence-avatar-size;
    height: $bdl-Presence-avatar-size;

    &:hover {
        cursor: pointer;
    }
}

.presence-overlay {
    width: 275px;
}
