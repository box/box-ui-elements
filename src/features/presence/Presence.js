import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage, injectIntl } from 'react-intl';
import classnames from 'classnames';

import { Flyout, Overlay } from '../../components/flyout';
import Tooltip from '../../components/tooltip';

// GROWTH-382
import Button from '../../components/button';

import { ARROW_DOWN, ENTER, SPACE } from '../../common/keyboard-events';

import PresenceDropdown from './PresenceDropdown';
import PresenceAvatar from './PresenceAvatar';
import { determineInteractionMessage } from './utils/presenceUtils';
import { collaboratorsPropType, flyoutPositionPropType } from './propTypes';

// GROWTH-382
import {
    GROWTH_382_EXPERIMENT_BUCKET,
    GROWTH_382_AUTOFLY_CLASS,
    GROWTH_382_AUTOFLY_CLASS_FIRST_LOAD,
} from './constants';
import messages from './messages';

import './Presence.scss';

class Presence extends Component {
    /* eslint-disable no-underscore-dangle */
    static propTypes = {
        /** Addtional attributes for avatar container */
        avatarAttributes: PropTypes.object,
        className: PropTypes.string,
        collaborators: PropTypes.arrayOf(collaboratorsPropType).isRequired,
        /** Addtional attributes for presence container */
        containerAttributes: PropTypes.object,
        /** Get Link callback */
        getLinkCallback: PropTypes.func,
        /** Invite button callback */
        inviteCallback: PropTypes.func,
        /** Maximum number of avatars to display before showing a +{n} avatar */
        maxDisplayedAvatars: PropTypes.number,
        /** Maximum number of collaborators before displaying a {maxAdditionalCollaboratorsNum}+ avatar */
        maxAdditionalCollaboratorsNum: PropTypes.number,
        /** Callback funtion for avatar mouseEnter, argument: id of user */
        onAvatarMouseEnter: PropTypes.func,
        /** Callback function for avatar mouseLeave */
        onAvatarMouseLeave: PropTypes.func,
        /** Callback funtion for Flyout events, argument: SyntheticEvent */
        onFlyoutClose: PropTypes.func,
        onFlyoutOpen: PropTypes.func,
        onFlyoutScroll: PropTypes.func,
        /** GROWTH-382 bucketing */
        experimentBucket: PropTypes.string,
        /** GROWTH-382 broadcast that the user wants to view stats from the flyout */
        onAccessStatsRequested: PropTypes.func,
        /** GROWTH-382 log that the user wants to view collaborators from the flyout */
        onClickViewCollaborators: PropTypes.func,
        /** Option to change the orientation of the dropdown. MUST be: bottom-right, bottom-left, bottom-center etc. or in this specific format */
        flyoutPosition: flyoutPositionPropType,
        /** Sets the tether constraint to scrollParent for the flyout */
        constrainToScrollParent: PropTypes.bool,
        /** Sets the tether constraint to window for the flyout */
        constrainToWindow: PropTypes.bool,
        /** Closes the flyout when window loses focus */
        closeOnWindowBlur: PropTypes.bool,
        intl: PropTypes.any,
    };

    static defaultProps = {
        className: '',
        maxDisplayedAvatars: 3,
        maxAdditionalCollaboratorsNum: 99,
        experimentBucket: null,
        flyoutPosition: 'bottom-left',
        constrainToScrollParent: true,
        constrainToWindow: false,
        closeOnWindowBlur: false,
    };

    state = {
        activeTooltip: null,
        isDropdownActive: false,
        showActivityPrompt: Boolean(
            this.props.collaborators.length &&
                this.props.onClickViewCollaborators &&
                this.props.experimentBucket === GROWTH_382_EXPERIMENT_BUCKET,
        ),
    };

    saveRefToContainer = el => {
        this.presenceContainerEl = el;
    };

    _showTooltip = id => {
        const { onAvatarMouseEnter } = this.props;
        this.setState({
            activeTooltip: id,
        });
        if (onAvatarMouseEnter) {
            onAvatarMouseEnter(id);
        }
    };

    _hideTooltip = () => {
        const { onAvatarMouseLeave } = this.props;
        this.setState({
            activeTooltip: null,
        });
        if (onAvatarMouseLeave) {
            onAvatarMouseLeave();
        }
    };

    _handleOverlayOpen = event => {
        const { onFlyoutOpen } = this.props;
        this.setState({
            isDropdownActive: true,
        });
        if (onFlyoutOpen) {
            onFlyoutOpen(event);
        }
    };

    _handleOverlayClose = event => {
        const { onFlyoutClose } = this.props;
        this.setState({
            isDropdownActive: false,
        });
        if (onFlyoutClose) {
            onFlyoutClose(event);
        }
    };

    stopPropagationAndPreventDefault = event => {
        event.stopPropagation();
        event.preventDefault();
    };

    openDropDown = () => {
        if (this.presenceContainerEl) {
            this.presenceContainerEl.click();
        }
    };

    handleKeyDown = event => {
        switch (event.key) {
            case ARROW_DOWN:
            case ENTER:
            case SPACE:
                this.openDropDown();
                this.stopPropagationAndPreventDefault(event);
                break;

            default:
                break;
        }
    };

    _renderTimestampMessage = (interactedAt, interactionType, isActive) => {
        const lastActionMessage = determineInteractionMessage(interactionType);
        const { intl } = this.props;
        const timeAgo = intl.formatRelativeTime
            ? intl.formatRelativeTime(interactedAt - Date.now())
            : intl.formatRelative(interactedAt);

        if (lastActionMessage) {
            return (
                <div>
                    <span className="presence-avatar-tooltip-event">
                        {isActive ? (
                            <FormattedMessage {...messages.activeNowText} />
                        ) : (
                            <FormattedMessage
                                {...lastActionMessage}
                                values={{
                                    timeAgo,
                                }}
                            />
                        )}
                    </span>
                </div>
            );
        }
        return null;
    };

    _renderAvatarTooltip = (name, interactedAt, interactionType, isActive) => (
        <div className="presence-avatar-tooltip-container">
            <div>
                <span className="presence-avatar-tooltip-name">{name}</span>
            </div>
            {this._renderTimestampMessage(interactedAt, interactionType, isActive)}
        </div>
    );

    // GROWTH-382 click through the first CTA, spawn the normal Presence dropdown
    _showRecentsFlyout = event => {
        const { onClickViewCollaborators } = this.props;
        onClickViewCollaborators();
        this.stopPropagationAndPreventDefault(event);
        this.setState({ showActivityPrompt: false });
    };

    render() {
        const {
            className,
            collaborators,
            maxDisplayedAvatars,
            maxAdditionalCollaboratorsNum,
            getLinkCallback,
            inviteCallback,
            onFlyoutScroll,
            avatarAttributes,
            containerAttributes,
            flyoutPosition,
            constrainToScrollParent,
            constrainToWindow,
            closeOnWindowBlur,
        } = this.props;

        const { activeTooltip, isDropdownActive } = this.state;
        const presenceCountClassName = classnames('presence-avatar', 'avatar', 'presence-count', {
            'dropdown-active': isDropdownActive,
        });

        // GROWTH-382
        const { experimentBucket, onAccessStatsRequested } = this.props;
        const { showActivityPrompt } = this.state;
        let requestAccessStats = null;

        if (!showActivityPrompt && experimentBucket === GROWTH_382_EXPERIMENT_BUCKET) {
            requestAccessStats = (
                // eslint-disable-next-line jsx-a11y/anchor-is-valid
                <a className="presence-dropdown-request-stats" href="#" onClick={onAccessStatsRequested}>
                    <FormattedMessage {...messages.previewPresenceFlyoutAccessStatsLink} />
                </a>
            );
        }

        const overlayClassNames = classnames('presence-dropdown-container', {
            [GROWTH_382_AUTOFLY_CLASS]: experimentBucket && !showActivityPrompt,
            [GROWTH_382_AUTOFLY_CLASS_FIRST_LOAD]: experimentBucket && showActivityPrompt,
        });

        const overlayContent = showActivityPrompt ? (
            <>
                <FormattedMessage {...messages.previewPresenceFlyoutCopy} />
                <Button className="btn-primary" onClick={this._showRecentsFlyout}>
                    <FormattedMessage {...messages.previewPresenceFlyoutActivityCTA} />
                </Button>
            </>
        ) : (
            <PresenceDropdown
                className="presence-dropdown"
                collaborators={collaborators}
                experimentBucket={experimentBucket}
                getLinkCallback={getLinkCallback}
                inviteCallback={inviteCallback}
                onScroll={onFlyoutScroll}
            />
        );
        return (
            <Flyout
                className={`presence ${className}`}
                closeOnWindowBlur={closeOnWindowBlur}
                constrainToScrollParent={constrainToScrollParent}
                constrainToWindow={constrainToWindow}
                isVisibleByDefault={showActivityPrompt}
                onClose={this._handleOverlayClose}
                onOpen={this._handleOverlayOpen}
                position={flyoutPosition}
            >
                {/* eslint-disable-next-line jsx-a11y/no-static-element-interactions */}
                <div
                    ref={this.saveRefToContainer}
                    className="presence-avatar-container"
                    onKeyDown={this.handleKeyDown}
                    {...containerAttributes}
                >
                    {collaborators.slice(0, maxDisplayedAvatars).map(collaborator => {
                        const { id, avatarUrl, name, isActive, interactedAt, interactionType } = collaborator;
                        return (
                            <Tooltip
                                key={id}
                                isShown={!isDropdownActive && activeTooltip === id}
                                position="bottom-center"
                                text={this._renderAvatarTooltip(name, interactedAt, interactionType, isActive)}
                            >
                                <PresenceAvatar
                                    avatarUrl={avatarUrl}
                                    id={id}
                                    isActive={isActive}
                                    name={name}
                                    onBlur={this._hideTooltip}
                                    onFocus={this._showTooltip.bind(this, id)}
                                    onMouseEnter={this._showTooltip.bind(this, id)}
                                    onMouseLeave={this._hideTooltip}
                                    {...avatarAttributes}
                                />
                            </Tooltip>
                        );
                    })}

                    {collaborators.length > maxDisplayedAvatars && (
                        // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
                        <div className={presenceCountClassName} tabIndex="0" {...avatarAttributes}>
                            {collaborators.length - maxDisplayedAvatars > maxAdditionalCollaboratorsNum
                                ? `${maxAdditionalCollaboratorsNum}+`
                                : `+${collaborators.length - maxDisplayedAvatars}`}
                        </div>
                    )}
                </div>
                <Overlay className={overlayClassNames} shouldDefaultFocus={false}>
                    {overlayContent}
                    {requestAccessStats}
                </Overlay>
            </Flyout>
        );
    }
}

export { Presence as PresenceComponent };
export default injectIntl(Presence);
