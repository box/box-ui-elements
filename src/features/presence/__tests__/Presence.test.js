/* eslint-disable no-underscore-dangle */
import React from 'react';
import { createIntl } from 'react-intl';
import collaboratorList from '../__mocks__/collaborators';
import { PresenceComponent as Presence } from '../Presence';

const GlobalDate = Date;

const intl = createIntl({});

describe('features/presence/Presence', () => {
    beforeEach(() => {
        global.Date = jest.fn(date => new GlobalDate(date));
        global.Date.now = () => 1000;
    });

    afterEach(() => {
        global.Date = GlobalDate;
    });
    describe('render()', () => {
        test('should correctly render empty state', () => {
            const collaborators = [];

            const wrapper = shallow(<Presence intl={intl} collaborators={collaborators} />);

            expect(wrapper.find('.presence-avatar-container').length).toBe(1);
            expect(wrapper.find('PresenceAvatar').length).toBe(0);
            expect(wrapper.find('.presence-count').length).toBe(0);
        });

        test('should set isDropdownActive to true and call OnFlyoutOpen when _handleOverlayOpen is called', () => {
            const onFlyoutOpenSpy = jest.fn();
            const wrapper = shallow(
                <Presence intl={intl} collaborators={collaboratorList} onFlyoutOpen={onFlyoutOpenSpy} />,
            );

            const instance = wrapper.instance();
            instance._handleOverlayOpen();

            expect(wrapper.state('isDropdownActive')).toBe(true);
            expect(onFlyoutOpenSpy).toHaveBeenCalled();
        });

        test('should set isDropdownActive to false and call OnFlyoutOpen and OnflyoutClose when _handleOverlayClose is called after _handleOverlayOpen', () => {
            const onFlyoutOpenSpy = jest.fn();
            const onFlyoutCloseSpy = jest.fn();
            const wrapper = shallow(
                <Presence
                    intl={intl}
                    collaborators={collaboratorList}
                    onFlyoutClose={onFlyoutCloseSpy}
                    onFlyoutOpen={onFlyoutOpenSpy}
                />,
            );

            const instance = wrapper.instance();
            instance._handleOverlayOpen();
            instance._handleOverlayClose();

            expect(wrapper.state('isDropdownActive')).toBe(false);
            expect(onFlyoutOpenSpy).toHaveBeenCalled();
            expect(onFlyoutCloseSpy).toHaveBeenCalled();
        });

        // GROWTH-382
        describe('Tests for presence autofly - GROWTH-382 AB test', () => {
            test('should render autofly on load when the experiment bucket is "flyout"', () => {
                const wrapper = shallow(
                    <Presence
                        intl={intl}
                        collaborators={collaboratorList}
                        experimentBucket="flyout"
                        onClickViewCollaborators={jest.fn()}
                    />,
                );
                expect(wrapper).toMatchSnapshot();
                expect(wrapper.find('.presence-autofly-first-load').length).toBe(1);
            });

            test('should not render autofly on load when the experiment bucket is "control" or null', () => {
                [
                    {
                        bucketName: 'control',
                    },
                    {
                        bucketName: null,
                    },
                ].forEach(({ bucketName }) => {
                    const wrapper = shallow(
                        <Presence
                            intl={intl}
                            collaborators={collaboratorList}
                            experimentBucket={bucketName}
                            onClickViewCollaborators={jest.fn()}
                        />,
                    );
                    expect(wrapper).toMatchSnapshot();
                    expect(wrapper.find('.presence-autofly-first-load').length).toBe(0);
                });
            });

            test('should not render autofly treatment when there are no collaborators to list', () => {
                const noCollaborators = [];
                const wrapper = shallow(
                    <Presence
                        intl={intl}
                        collaborators={noCollaborators}
                        experimentBucket="flyout"
                        onClickViewCollaborators={jest.fn()}
                    />,
                );
                expect(wrapper).toMatchSnapshot();
                expect(wrapper.find('.presence-autofly-first-load').length).toBe(0);
            });

            test('should not render autofly treatment when there is no click handler to open the recents panel', () => {
                const wrapper = shallow(
                    <Presence intl={intl} collaborators={collaboratorList} experimentBucket="flyout" />,
                );
                expect(wrapper).toMatchSnapshot();
                expect(wrapper.find('.presence-autofly-first-load').length).toBe(0);
            });

            test('_showRecentsFlyout performs correct actions', () => {
                const mockOnClickViewCollaborators = jest.fn();
                const mockEvent = {
                    preventDefault: () => {},
                    stopPropagation: () => {},
                };
                const wrapper = shallow(
                    <Presence
                        intl={intl}
                        collaborators={collaboratorList}
                        experimentBucket="flyout"
                        onClickViewCollaborators={mockOnClickViewCollaborators}
                    />,
                );

                wrapper.instance()._showRecentsFlyout(mockEvent);
                expect(mockOnClickViewCollaborators.mock.calls.length).toBe(1);
                expect(wrapper.state('showActivityPrompt')).toBe(false);
            });

            test('should show the Presence dropdown when state.showActivityPrompt is false', () => {
                const wrapper = shallow(
                    <Presence
                        intl={intl}
                        collaborators={collaboratorList}
                        experimentBucket="flyout"
                        onClickViewCollaborators={jest.fn()}
                    />,
                );
                expect(wrapper.find('PresenceCollaboratorsList').length).toBe(0);
                wrapper.setState({ showActivityPrompt: false }).update();
                expect(wrapper.find('PresenceCollaboratorsList').length).toBe(1);
                expect(wrapper.find('.presence-overlay-request-stats').length).toBe(1);
            });

            test('calls callback to open access stats when that link is clicked', () => {
                const mockRequestAccessStats = jest.fn();
                const wrapper = shallow(
                    <Presence
                        intl={intl}
                        collaborators={collaboratorList}
                        experimentBucket="flyout"
                        onAccessStatsRequested={mockRequestAccessStats}
                        onClickViewCollaborators={jest.fn()}
                    />,
                );
                wrapper.setState({ showActivityPrompt: false }).update();

                wrapper.find('.presence-overlay-request-stats').simulate('click');

                expect(mockRequestAccessStats.mock.calls.length).toBe(1);
            });
        });
        // end GROWTH-382
    });
});
