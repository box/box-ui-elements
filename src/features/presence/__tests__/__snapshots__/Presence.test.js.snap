// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features/presence/Presence render() Tests for presence autofly - GROWTH-382 AB test should not render autofly on load when the experiment bucket is "control" or null 1`] = `
<Flyout
  className="presence "
  closeOnClick={true}
  closeOnClickOutside={true}
  closeOnWindowBlur={false}
  constrainToScrollParent={true}
  constrainToWindow={false}
  isVisibleByDefault={false}
  onClose={[Function]}
  onOpen={[Function]}
  openOnHover={false}
  openOnHoverDelayTimeout={300}
  portaledClasses={Array []}
  position="bottom-left"
>
  <ForwardRef(PresenceAvatarList)
    aria-label="Recent Activity"
    className="presence-avatar-container"
    collaborators={
      Array [
        Object {
          "avatarUrl": "",
          "id": "1",
          "interactedAt": 999,
          "interactionType": "user.item_preview",
          "isActive": false,
          "name": "e",
        },
        Object {
          "avatarUrl": "",
          "id": "2",
          "interactedAt": 999,
          "interactionType": "user.item_upload",
          "isActive": false,
          "name": "d",
        },
        Object {
          "avatarUrl": "",
          "id": "3",
          "interactedAt": 999,
          "interactionType": "user.comment_create",
          "isActive": false,
          "name": "c",
        },
        Object {
          "avatarUrl": "",
          "id": "4",
          "interactedAt": 999,
          "interactionType": "user.item_preview",
          "isActive": false,
          "name": "b",
        },
        Object {
          "avatarUrl": "",
          "id": "5",
          "interactedAt": 999,
          "interactionType": "user.item_upload",
          "isActive": false,
          "name": "a",
        },
      ]
    }
    hideTooltips={false}
    maxAdditionalCollaborators={99}
    maxDisplayedAvatars={3}
    onKeyDown={[Function]}
  />
  <Overlay
    className="presence-overlay presence-autofly"
    shouldDefaultFocus={false}
  >
    <PresenceCollaboratorsList
      collaborators={
        Array [
          Object {
            "avatarUrl": "",
            "id": "1",
            "interactedAt": 999,
            "interactionType": "user.item_preview",
            "isActive": false,
            "name": "e",
          },
          Object {
            "avatarUrl": "",
            "id": "2",
            "interactedAt": 999,
            "interactionType": "user.item_upload",
            "isActive": false,
            "name": "d",
          },
          Object {
            "avatarUrl": "",
            "id": "3",
            "interactedAt": 999,
            "interactionType": "user.comment_create",
            "isActive": false,
            "name": "c",
          },
          Object {
            "avatarUrl": "",
            "id": "4",
            "interactedAt": 999,
            "interactionType": "user.item_preview",
            "isActive": false,
            "name": "b",
          },
          Object {
            "avatarUrl": "",
            "id": "5",
            "interactedAt": 999,
            "interactionType": "user.item_upload",
            "isActive": false,
            "name": "a",
          },
        ]
      }
      experimentBucket="control"
    />
  </Overlay>
</Flyout>
`;

exports[`features/presence/Presence render() Tests for presence autofly - GROWTH-382 AB test should not render autofly on load when the experiment bucket is "control" or null 2`] = `
<Flyout
  className="presence "
  closeOnClick={true}
  closeOnClickOutside={true}
  closeOnWindowBlur={false}
  constrainToScrollParent={true}
  constrainToWindow={false}
  isVisibleByDefault={false}
  onClose={[Function]}
  onOpen={[Function]}
  openOnHover={false}
  openOnHoverDelayTimeout={300}
  portaledClasses={Array []}
  position="bottom-left"
>
  <ForwardRef(PresenceAvatarList)
    aria-label="Recent Activity"
    className="presence-avatar-container"
    collaborators={
      Array [
        Object {
          "avatarUrl": "",
          "id": "1",
          "interactedAt": 999,
          "interactionType": "user.item_preview",
          "isActive": false,
          "name": "e",
        },
        Object {
          "avatarUrl": "",
          "id": "2",
          "interactedAt": 999,
          "interactionType": "user.item_upload",
          "isActive": false,
          "name": "d",
        },
        Object {
          "avatarUrl": "",
          "id": "3",
          "interactedAt": 999,
          "interactionType": "user.comment_create",
          "isActive": false,
          "name": "c",
        },
        Object {
          "avatarUrl": "",
          "id": "4",
          "interactedAt": 999,
          "interactionType": "user.item_preview",
          "isActive": false,
          "name": "b",
        },
        Object {
          "avatarUrl": "",
          "id": "5",
          "interactedAt": 999,
          "interactionType": "user.item_upload",
          "isActive": false,
          "name": "a",
        },
      ]
    }
    hideTooltips={false}
    maxAdditionalCollaborators={99}
    maxDisplayedAvatars={3}
    onKeyDown={[Function]}
  />
  <Overlay
    className="presence-overlay"
    shouldDefaultFocus={false}
  >
    <PresenceCollaboratorsList
      collaborators={
        Array [
          Object {
            "avatarUrl": "",
            "id": "1",
            "interactedAt": 999,
            "interactionType": "user.item_preview",
            "isActive": false,
            "name": "e",
          },
          Object {
            "avatarUrl": "",
            "id": "2",
            "interactedAt": 999,
            "interactionType": "user.item_upload",
            "isActive": false,
            "name": "d",
          },
          Object {
            "avatarUrl": "",
            "id": "3",
            "interactedAt": 999,
            "interactionType": "user.comment_create",
            "isActive": false,
            "name": "c",
          },
          Object {
            "avatarUrl": "",
            "id": "4",
            "interactedAt": 999,
            "interactionType": "user.item_preview",
            "isActive": false,
            "name": "b",
          },
          Object {
            "avatarUrl": "",
            "id": "5",
            "interactedAt": 999,
            "interactionType": "user.item_upload",
            "isActive": false,
            "name": "a",
          },
        ]
      }
      experimentBucket={null}
    />
  </Overlay>
</Flyout>
`;

exports[`features/presence/Presence render() Tests for presence autofly - GROWTH-382 AB test should not render autofly treatment when there are no collaborators to list 1`] = `
<Flyout
  className="presence "
  closeOnClick={true}
  closeOnClickOutside={true}
  closeOnWindowBlur={false}
  constrainToScrollParent={true}
  constrainToWindow={false}
  isVisibleByDefault={false}
  onClose={[Function]}
  onOpen={[Function]}
  openOnHover={false}
  openOnHoverDelayTimeout={300}
  portaledClasses={Array []}
  position="bottom-left"
>
  <ForwardRef(PresenceAvatarList)
    aria-label="Recent Activity"
    className="presence-avatar-container"
    collaborators={Array []}
    hideTooltips={false}
    maxAdditionalCollaborators={99}
    maxDisplayedAvatars={3}
    onKeyDown={[Function]}
  />
  <Overlay
    className="presence-overlay presence-autofly"
    shouldDefaultFocus={false}
  >
    <PresenceCollaboratorsList
      collaborators={Array []}
      experimentBucket="flyout"
    />
    <a
      className="presence-overlay-request-stats"
      href="#"
    >
      <FormattedMessage
        defaultMessage="See all activity"
        id="boxui.presence.previewPresenceFlyoutAccessStatsLink"
      />
    </a>
  </Overlay>
</Flyout>
`;

exports[`features/presence/Presence render() Tests for presence autofly - GROWTH-382 AB test should not render autofly treatment when there is no click handler to open the recents panel 1`] = `
<Flyout
  className="presence "
  closeOnClick={true}
  closeOnClickOutside={true}
  closeOnWindowBlur={false}
  constrainToScrollParent={true}
  constrainToWindow={false}
  isVisibleByDefault={false}
  onClose={[Function]}
  onOpen={[Function]}
  openOnHover={false}
  openOnHoverDelayTimeout={300}
  portaledClasses={Array []}
  position="bottom-left"
>
  <ForwardRef(PresenceAvatarList)
    aria-label="Recent Activity"
    className="presence-avatar-container"
    collaborators={
      Array [
        Object {
          "avatarUrl": "",
          "id": "1",
          "interactedAt": 999,
          "interactionType": "user.item_preview",
          "isActive": false,
          "name": "e",
        },
        Object {
          "avatarUrl": "",
          "id": "2",
          "interactedAt": 999,
          "interactionType": "user.item_upload",
          "isActive": false,
          "name": "d",
        },
        Object {
          "avatarUrl": "",
          "id": "3",
          "interactedAt": 999,
          "interactionType": "user.comment_create",
          "isActive": false,
          "name": "c",
        },
        Object {
          "avatarUrl": "",
          "id": "4",
          "interactedAt": 999,
          "interactionType": "user.item_preview",
          "isActive": false,
          "name": "b",
        },
        Object {
          "avatarUrl": "",
          "id": "5",
          "interactedAt": 999,
          "interactionType": "user.item_upload",
          "isActive": false,
          "name": "a",
        },
      ]
    }
    hideTooltips={false}
    maxAdditionalCollaborators={99}
    maxDisplayedAvatars={3}
    onKeyDown={[Function]}
  />
  <Overlay
    className="presence-overlay presence-autofly"
    shouldDefaultFocus={false}
  >
    <PresenceCollaboratorsList
      collaborators={
        Array [
          Object {
            "avatarUrl": "",
            "id": "1",
            "interactedAt": 999,
            "interactionType": "user.item_preview",
            "isActive": false,
            "name": "e",
          },
          Object {
            "avatarUrl": "",
            "id": "2",
            "interactedAt": 999,
            "interactionType": "user.item_upload",
            "isActive": false,
            "name": "d",
          },
          Object {
            "avatarUrl": "",
            "id": "3",
            "interactedAt": 999,
            "interactionType": "user.comment_create",
            "isActive": false,
            "name": "c",
          },
          Object {
            "avatarUrl": "",
            "id": "4",
            "interactedAt": 999,
            "interactionType": "user.item_preview",
            "isActive": false,
            "name": "b",
          },
          Object {
            "avatarUrl": "",
            "id": "5",
            "interactedAt": 999,
            "interactionType": "user.item_upload",
            "isActive": false,
            "name": "a",
          },
        ]
      }
      experimentBucket="flyout"
    />
    <a
      className="presence-overlay-request-stats"
      href="#"
    >
      <FormattedMessage
        defaultMessage="See all activity"
        id="boxui.presence.previewPresenceFlyoutAccessStatsLink"
      />
    </a>
  </Overlay>
</Flyout>
`;

exports[`features/presence/Presence render() Tests for presence autofly - GROWTH-382 AB test should render autofly on load when the experiment bucket is "flyout" 1`] = `
<Flyout
  className="presence "
  closeOnClick={true}
  closeOnClickOutside={true}
  closeOnWindowBlur={false}
  constrainToScrollParent={true}
  constrainToWindow={false}
  isVisibleByDefault={true}
  onClose={[Function]}
  onOpen={[Function]}
  openOnHover={false}
  openOnHoverDelayTimeout={300}
  portaledClasses={Array []}
  position="bottom-left"
>
  <ForwardRef(PresenceAvatarList)
    aria-label="Recent Activity"
    className="presence-avatar-container"
    collaborators={
      Array [
        Object {
          "avatarUrl": "",
          "id": "1",
          "interactedAt": 999,
          "interactionType": "user.item_preview",
          "isActive": false,
          "name": "e",
        },
        Object {
          "avatarUrl": "",
          "id": "2",
          "interactedAt": 999,
          "interactionType": "user.item_upload",
          "isActive": false,
          "name": "d",
        },
        Object {
          "avatarUrl": "",
          "id": "3",
          "interactedAt": 999,
          "interactionType": "user.comment_create",
          "isActive": false,
          "name": "c",
        },
        Object {
          "avatarUrl": "",
          "id": "4",
          "interactedAt": 999,
          "interactionType": "user.item_preview",
          "isActive": false,
          "name": "b",
        },
        Object {
          "avatarUrl": "",
          "id": "5",
          "interactedAt": 999,
          "interactionType": "user.item_upload",
          "isActive": false,
          "name": "a",
        },
      ]
    }
    hideTooltips={false}
    maxAdditionalCollaborators={99}
    maxDisplayedAvatars={3}
    onKeyDown={[Function]}
  />
  <Overlay
    className="presence-overlay presence-autofly-first-load"
    shouldDefaultFocus={false}
  >
    <FormattedMessage
      defaultMessage="Quickly see who has commented on, edited, or viewed this file."
      id="boxui.presence.previewPresenceFlyoutCopy"
    />
    <Button
      className="btn-primary"
      isLoading={false}
      onClick={[Function]}
      showRadar={false}
      type="submit"
    >
      <FormattedMessage
        defaultMessage="View Recent Activity"
        id="boxui.presence.previewPresenceFlyoutActivityCTA"
      />
    </Button>
  </Overlay>
</Flyout>
`;
